File: src/linkedin/captcha.ts
import type { Page } from 'playwright';

// URL path patterns that indicate LinkedIn security challenge
const CAPTCHA_URL_PATTERNS = [
  '/checkpoint/',
  '/captcha',
  '/uas/login',
  '/security',
  '/challenge/',
];

// DOM selectors that indicate a captcha/challenge is present
const CAPTCHA_SELECTORS = [
  'form[id="captcha-challenge"]',
  '[data-test-id="captcha"]',
  '[data-challenge-id]',
];

// Page title substrings (case-insensitive)
const CAPTCHA_TITLE_PATTERNS = [
  'security verification',
  'security check',
  'confirm',
  'challenge',
  'verify',
];

export async function isCaptchaPage(page: Page): Promise<boolean> {
  try {
    const url = page.url().toLowerCase();
    if (CAPTCHA_URL_PATTERNS.some(p => url.includes(p))) return true;

    const title = (await page.title()).toLowerCase();
    if (CAPTCHA_TITLE_PATTERNS.some(p => title.includes(p))) return true;

    for (const sel of CAPTCHA_SELECTORS) {
      const el = await page.$(sel);
      if (el) return true;
    }

    return false;
  } catch {
    return false;
  }
}

export class CaptchaError extends Error {
  constructor(url: string) {
    super(`Captcha/challenge detected at ${url} — stopping session`);
    this.name = 'CaptchaError';
  }
}

// For testing — exported URL pattern list
export const captchaUrlPatterns = CAPTCHA_URL_PATTERNS;
export const captchaSelectors = CAPTCHA_SELECTORS;
export const captchaTitlePatterns = CAPTCHA_TITLE_PATTERNS;

File: crates/linkedin-ops/src/captcha.rs
use thiserror::Error;

#[derive(Debug, Error)]
#[error("captcha/challenge detected — stopping all actions. Run `forge-linkedin status` and check for alerts.")]
pub struct CaptchaError;

/// Detect captcha / security challenge from page signals.
/// Call this after every navigation.
pub fn detect_captcha(url: &str, page_title: &str, html_snippet: &str) -> bool {
    let url_lc = url.to_lowercase();
    let title_lc = page_title.to_lowercase();
    let html_lc = html_snippet.to_lowercase();

    // URL-based signals
    if url_lc.contains("/checkpoint/challenge") {
        return true;
    }
    if url_lc.contains("/uas/login") && url_lc.contains("challenge") {
        return true;
    }
    if url_lc.contains("/checkpoint/rm/") {
        return true;
    }

    // Title-based signals
    if title_lc.contains("security verification") {
        return true;
    }
    if title_lc.contains("help us confirm") {
        return true;
    }
    if title_lc.contains("let's do a quick security check") {
        return true;
    }

    // HTML-based signals
    if html_lc.contains(r#"id="captcha-challenge""#) {
        return true;
    }
    if html_lc.contains("data-test-id=\"captcha\"") {
        return true;
    }

    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_clean_page_no_captcha() {
        assert!(!detect_captcha(
            "https://www.linkedin.com/feed/",
            "LinkedIn",
            "<div class='feed'></div>"
        ));
    }

    #[test]
    fn test_checkpoint_url() {
        assert!(detect_captcha(
            "https://www.linkedin.com/checkpoint/challenge/abc123",
            "LinkedIn",
            ""
        ));
    }

    #[test]
    fn test_security_verification_title() {
        assert!(detect_captcha(
            "https://www.linkedin.com/feed/",
            "Security Verification — LinkedIn",
            ""
        ));
    }

    #[test]
    fn test_help_confirm_title() {
        assert!(detect_captcha(
            "https://www.linkedin.com/feed/",
            "Help us confirm you're human",
            ""
        ));
    }

    #[test]
    fn test_captcha_form_in_html() {
        assert!(detect_captcha(
            "https://www.linkedin.com/feed/",
            "LinkedIn",
            r#"<form id="captcha-challenge"><input type="text"></form>"#
        ));
    }

    #[test]
    fn test_case_insensitive() {
        assert!(detect_captcha(
            "https://www.linkedin.com/Checkpoint/Challenge/xyz",
            "SECURITY VERIFICATION",
            ""
        ));
    }
}
